File: stress/stress/thermal.py
"""SP-2: Thermal Cycling — periodic stress modulation.

Modulates SP-1 fault rate using a sinusoidal function.
Purely deterministic; no RNG needed.
"""
from __future__ import annotations

import math
import random
from dataclasses import dataclass

from stress.stress.base import Stressor, StressorConfig


@dataclass(frozen=True)
class ThermalConfig(StressorConfig):
    """SP-2 configuration."""
    cycle_period_s: float = 600.0  # Period of one thermal cycle in seconds
    amplitude: float = 3.0  # Peak multiplier on SP-1 fault rate


class ThermalStressor(Stressor):
    """Deterministic periodic thermal cycling.

    Produces a multiplier >= 1.0 that modulates the SP-1 fault rate.
    """

    def __init__(self, config: ThermalConfig):
        super().__init__(config)
        self._cfg: ThermalConfig = config

    def reset(self) -> None:
        self._rng = random.Random(self._config.seed)
        self._active = False
        self._start_time = None

    def get_thermal_multiplier(self, t: float) -> float:
        """Get the thermal multiplier at time t.

        Returns a value >= 1.0 that should be applied to SP-1's fault rate.
        The multiplier follows a sinusoidal curve:
          1.0 + (amplitude - 1.0) * (0.5 + 0.5 * sin(2*pi*t / period))
        """
        if not self._active or not self._cfg.enabled:
            return 1.0

        elapsed = self._elapsed(t)
        phase = 2.0 * math.pi * elapsed / self._cfg.cycle_period_s
        # Range: [1.0, amplitude]
        return 1.0 + (self._cfg.amplitude - 1.0) * (0.5 + 0.5 * math.sin(phase))

File: stress-ref/src/stress/thermal.rs
use std::f64::consts::PI;

/// SP-2: Thermal Cycling — deterministic periodic multiplier on SP-1 fault rate.
pub struct ThermalStressor {
    enabled: bool,
    cycle_period_s: f64,
    amplitude: f64,
    active: bool,
    start_time: Option<f64>,
}

impl ThermalStressor {
    pub fn new(enabled: bool, _seed: u64, cycle_period_s: f64, amplitude: f64) -> Self {
        Self {
            enabled,
            cycle_period_s,
            amplitude: amplitude.max(1.0), // Thermal stress can only increase fault rate
            active: false,
            start_time: None,
        }
    }

    pub fn start(&mut self, t: f64) {
        self.active = true;
        self.start_time = Some(t);
    }

    pub fn stop(&mut self) {
        self.active = false;
    }

    pub fn enabled(&self) -> bool {
        self.enabled
    }

    /// Returns thermal multiplier >= 1.0 following sinusoidal curve.
    pub fn get_thermal_multiplier(&self, t: f64) -> f64 {
        if !self.active || !self.enabled {
            return 1.0;
        }
        let elapsed = t - self.start_time.unwrap_or(t);
        let phase = 2.0 * PI * elapsed / self.cycle_period_s;
        1.0 + (self.amplitude - 1.0) * (0.5 + 0.5 * phase.sin())
    }
}
